### PR TITLE
ARROW-10842 [Rust] decouple IO from json reader, fix crash during json schema inference with invalid json

### PR DESCRIPTION
**Pull request by houqp**
Other related changes:

 * reuse the same IO code between schema inference and record batch reader.
 * reuse string buffer to avoid repeatitive memory allocation
 * ignore new lines when reading json records
 * avoid hard crash during JSON schema inference caused by invalid json

### CODE REVIEW

File: rust/arrow/src/json/reader.rs [hunk 1]
```diff
@@ -250,19 +326,17 @@ pub fn infer_json_schema<R: Read>(
...
 ) -> Result<SchemaRef> {
-    let mut values: HashMap<String, HashSet<DataType>> = HashMap::new();
-
-    let mut line = String::new();
-    for _ in 0..max_read_records.unwrap_or(std::usize::MAX) {
-        reader.read_line(&mut line)?;
-        if line.is_empty() {
-            break;
-        }
-        let record: Value = serde_json::from_str(&line.trim()).expect("Not valid JSON");
```
Comment (alamb) [inline rust/arrow/src/json/reader.rs:261]: is this the line that causes a hard crash on invalid json?

**Review by alamb — COMMENTED** [review]
Thanks @houqp  -- I think this code looks really nice and is a substantial improvement. 

I wonder what you think about adding some additional tests for the new behavior?  Specifically: 

1. Invalid JSON
2. Empty lines (and the correct record count)

Your code looks correct on inspection, so I think the primary value of such tests would be to encode the expected behavior so that future changes / refactors wouldn't cause a regression.

File: rust/arrow/src/json/reader.rs (hunk 1 above)
Comment (houqp) [reply rust/arrow/src/json/reader.rs:261]: that's right.

**Comment by houqp** [comment]
Thanks @alamb for the review, I will add more tests as well as also change to decouple IO from record batch reader.

**Review by alamb — APPROVED** [review]
Looks good to me. Thank you @houqp 

cc @nevi-me

**Comment by alamb** [comment]
@houqp  it seems as though there are some clippy errors that are preventing a clean test run: https://github.com/apache/arrow/pull/8867/checks?check_run_id=1522963773


```
 Compiling arrow v3.0.0-SNAPSHOT (/__w/arrow/arrow/rust/arrow)
error: explicit lifetimes given in parameter types where they could be elided (or replaced with `'_` if needed by type declaration)
   --> arrow/src/json/reader.rs:546:5
    |
546 | /     pub fn next_batch<'a, 'b, I>(
547 | |         &'a self,
548 | |         value_iter: &'b mut I,
549 | |     ) -> Result<Option<RecordBatch>>
    | |____________________________________^
    |
    = note: `-D clippy::needless-lifetimes` implied by `-D warnings`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes

error: explicit lifetimes given in parameter types where they could be elided (or replaced with `'_` if needed by type declaration)
   --> arrow/src/json/reader.rs:546:5
    |
546 | /     pub fn next_batch<'a, 'b, I>(
547 | |         &'a self,
548 | |         value_iter: &'b mut I,
549 | |     ) -> Result<Option<RecordBatch>>
    | |____________________________________^
    |
    = note: `-D clippy::needless-lifetimes` implied by `-D warnings`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes

error: aborting due to previous error

error: could not compile `arrow`
```

**Comment by houqp** [comment]
my bad, fixed clippy error

**Pull request closed without merging**